keps/sig-cluster-lifecycle/kubeadm/20190916-kubeadm-operator.md

```diff
@@ -0,0 +1,202 @@
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed](#infrastructure-needed)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [x] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location
+      in kubernetes/enhancements, not the initial KEP PR)
+- [x] KEP approvers have set the KEP status to `implementable`
+- [x] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings,
+      relevant PRs/issues, release notes
+
+## Summary
+
+Kubeadm operator would like to enable declarative control of kubeadm workflows, automating the execution and the
+orchestration of such tasks across existing nodes in a cluster.
+
+## Motivation
+
+Kubeadm binary can execute operations only on the machine where it is running e.g. it is not possible to execute
+operations on other nodes, to copy files across nodes, etc.
+
+As a consequence, most of the kubeadm workflows, like [kubeadm upgrade](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-upgrade/),
+consists of a complex sequence of tasks that should be manually executed and orchestrated across all the existing nodes
+in the cluster.
+
+Such a user experience is not ideal due to the error-prone nature of humans running commands. The manual approach
+can be considered a blocker for implementing more complex workflows such as rotating certificate authorities,
+modifying the settings of an existing cluster or any task that requires coordination of more than one Kubernetes node.
+
+This KEP aims to address such problems by applying the operator pattern to kubeadm workflows. 
+
+### Goals
+
+- To allow declarative control of kubeadm operations that lead to "in place" mutations[1] of kubeadm generated
+  artifacts. More specifically kubeadm artifacts are static pod manifests, certificates, kubeconfig file,
+  bootstrap token, kubeadm generated configmap and secrets, while this proposal initially includes the following
+  kubeadm workflows:
+  - kubeadm upgrade
+  - certificate renewal
+  - certificate authority rotation (NEW)
+  - change configuration in an existing cluster (NEW) 
+
+> [1] Please note that we are referring to "in place" mutations of kubeadm generated artifacts in order to highlight
```

Comment:
CAPI recently merged tighter integration w/ cert-manager - I'm wondering if this footnote is outdated?

Reply:
At first sight, I don't see relations between cert-manager and the kubeadm-operator, but I prefer to double-check. Can you provide a link?